nsg-netbox.py

```diff
@@ -50,87 +90,311 @@ def run(self):
+                           }
+            self.log.info(f'NetSpyGlass: {len(nsg_devices):>6} devices')
 
-            nsg_devices = nsg.get_devices()
-            self.log.info('NetSpyGlass: {0} devices'.format(len(nsg_devices)))
+            tags_ = self.nsg.get_device_tags()
+            tags = {}
+            for i in tags_:
+                key = i.get("device", {}).get("address")
+                if not tags.get(key):
+                    tags[key] = {"id": i.get("device", {}).get("id"),
+                                 "tags": dict(t.split(".", maxsplit=1) for t in i.get("tags", {}))}
+                else:
+                    tags[key]["tags"].update(dict(t.split(".", maxsplit=1) for t in i.get("tags", {})))
 
             nb_dev_set = set(nbox_devices.keys())
             nsg_dev_set = set(nsg_devices.keys())
             to_add = set.difference(nb_dev_set, nsg_dev_set)  # set of addresses as strings
             to_remove = set.difference(nsg_dev_set, nb_dev_set)  # set of addresses as strings
 
+            if to_remove:
+                self.log.info('DELETE devices: {0}'.format(to_remove))
+                self.nsg.delete_devices(list(nsg_devices[addr]['id'] for addr in to_remove))
+                wait_tasks_done()
+                remove_tags = make_remove_tag_dict(nsg_devices=to_remove, nsg_tags=tags)
+                if remove_tags:
+                    for i in remove_tags:
+                        self.log.info(f" tags list to delete: {i}")
+                    self.nsg.post_device_tags(tag_list=remove_tags, operation="DELETE")
+
             if to_add:
+                payload = list(self.make_add_device_dict(addr, nbox_devices[addr]) for addr in to_add)
                 self.log.info('ADD devices:    {0}'.format(to_add))
-                nsg.add_devices(list(self.make_add_device_dict(addr, nbox_devices[addr]) for addr in to_add))
+                self.nsg.add_devices(payload)
 
-            if to_remove:
-                self.log.info('DELETE devices: {0}'.format(to_remove))
-                nsg.delete_devices(list(nsg_devices[addr]['id'] for addr in to_remove))
+            tag_list = make_add_tag_dict(nbox_devices=nbox_devices,
+                                         nsg_tags=tags)
+            # wait for devices added to nsg
+            wait_tasks_done()
+
+            if tag_list:
+                for i in tag_list:
+                    self.log.info(f" tags list to update: {i}")
+                self.nsg.post_device_tags(tag_list=tag_list)
+            else:
+                self.log.info(f" no devices with new/changed tags found")
 
         except pynetbox.core.query.RequestError as e:
             self.log.error('Netbox API call has failed: {0}'.format(e))
         except Exception as e:
             self.log.exception('Unknown exception: %s', e)
+        # schedule next run
+        self.scheduler.enter(delay=self.interval_sec, priority=1, action=self.run)
```

Comment:
the original implementation called `self.scheduler.enter()` at the beginning, but you removed that call. Now you have to make sure you call it everywhere you exit this function, which is two places instead of one. This change would be justified if there was a code path when we do not need to schedule next run, but such code path does not exist. We have to always schedule next run, so we might as well do it at the entry to this function

Reply:
moved to the beginning